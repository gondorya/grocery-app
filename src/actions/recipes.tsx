import uuid from 'uuid';

const addRecipe = (
	{ title = '', description = '', steps = [], ingredients = [], date = new Date(), img = '' } = {}
) => {
	return {
		type: 'ADD_RECIPE',
		recipe: {
			id: uuid(),
			title,
			description,
			steps,
			ingredients,
			img,
			date
		}
	};
};

const removeRecipe = ({ id = '' } = {}) => {
	return {
		type: 'REMOVE_RECIPE',
		id
	};
};

const editRecipe = (id, updates) => {
	return {
		type: 'EDIT_RECIPE',
		id,
		updates
	};
};

export { addRecipe, removeRecipe, editRecipe };
