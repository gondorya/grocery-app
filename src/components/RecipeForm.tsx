import React from 'react';
import NewStep from './atoms/NewStep';
import Button from './atoms/Button';

type Props = {
	stepsInputs: Array<string>;
	title: string;
	description: string;
	img: string;
	date: number;
	error: boolean | string;
	steps: Array<string>;
};

class RecipeForm extends React.Component<{ newRecipe: any }, Props> {
	constructor(props) {
		super(props);
	}

	state = {
		stepsInputs: [ 'step-0' ],
		title: '',
		description: '',
		img: '',
		date: 0,
		error: false,
		steps: []
	};

	addNewStep = () => {
		const keyValue = `step-${this.state.stepsInputs.length}`;
		this.setState((prevState) => ({
			stepsInputs: prevState.stepsInputs.concat([ keyValue ])
		}));
	};

	removeStep = (stepName) => {
		const result = this.state.stepsInputs.filter((step) => {
			return step !== stepName;
		});

		this.setState(() => ({
			stepsInputs: result
		}));
	};

	submitHandler = (e) => {
		e.preventDefault();
		this.setState(() => ({ error: false }));
		this.props.newRecipe({
			title: this.state.title,
			description: this.state.description,
			img: this.state.img,
			steps: this.state.steps
		});
	};

	onTitleChange = (e) => {
		const title = e.target.value;
		this.setState(() => ({ title }));
	};

	onDescriptionChange = (e) => {
		const description = e.target.value;
		this.setState(() => ({ description }));
	};

	onImgChange = (e) => {
		const img = e.target.value;
		this.setState(() => ({ img }));
	};

	onStepChange = (index, value) => {
		this.setState((prevState) => {
			const list = prevState.steps;
			list[index] = value;
			return {
				steps: list
			};
		});
	};

	render() {
		return (
			<div>
				<h2>Add New Recipe</h2>
				<form className="form" onSubmit={this.submitHandler}>
					<label>
						Title
						<input
							type="text"
							name="title"
							className="form__input form__input--big"
							onChange={this.onTitleChange}
						/>
					</label>
					<label>
						Description
						<textarea name="description" className="form__input form__input--big" />
					</label>
					<label>
						Photo
						<input type="file" name="photo" className="form__input" />
					</label>
					<div className="form__group">
						{this.state.stepsInputs.map((step, index) => (
							<NewStep
								key={step}
								value={step}
								buttonHandler={this.removeStep}
								onChange={(e) => this.onStepChange(index, e.target.value)}
							/>
						))}
						<Button text="Add New Step" type="button" className="btn--primary" onClick={this.addNewStep} />
					</div>
					<button type="submit" className="btn btn--big btn--submit">
						Create New Recipe
					</button>
				</form>
			</div>
		);
	}
}

export default RecipeForm;
