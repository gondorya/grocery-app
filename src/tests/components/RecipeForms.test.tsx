import React from 'react';
import { shallow } from 'enzyme';
import RecipeForm from '../../components/RecipeForm';
import recipes from '../fixtures/recipes';

test('Recipe form without data is rendering', () => {
	const wrapper = shallow(<RecipeForm />);
	expect(wrapper).toMatchSnapshot();
});

test('Recipe form with data is rendering', () => {
	const wrapper = shallow(<RecipeForm recipe={recipes[0]} />);
	expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid form submission', () => {
	const wrapper = shallow(<RecipeForm />);
	expect(wrapper.state('error')).toBe(false);
	wrapper.find('form').simulate('submit', { preventDefault: () => {} });
	expect(wrapper.state('error').length).toBeGreaterThan(0);
});

test('should set descritpion change', () => {
	const value = 'i change title';
	const wrapper = shallow(<RecipeForm />);
	wrapper.find('input').at(1).simulate('change', {
		target: { value }
	});
	expect(wrapper.state('title')).toBe(value);
});

test('should call onSubmit prop for valid form submission', () => {
	const onSubmitSpy = jest.fn();
	const wrapper = shallow(<RecipeForm newRecipe={onSubmitSpy} recipe={recipes[0]} />);
	wrapper.find('form').simulate('submit', { preventDefault: () => {} });
	expect(wrapper.state('error')).toBe(false);
	expect(onSubmitSpy).toHaveBeenLastCalledWith({
		title: 'fish and chips',
		description: 'english street food',
		date: '2018-09-15T00:00:00+02:00',
		ingredients: [ 'fish', 'chips', 'peas' ],
		img: '',
		steps: [ 'step 1', 'step 2', 'step 3' ]
	});
});
