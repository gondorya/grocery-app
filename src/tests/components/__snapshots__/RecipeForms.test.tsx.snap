// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recipe form with data is rendering 1`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <input
      type="date"
      value="2018-09-15T00:00:00+02:00"
    />
    <label>
      Title
      <input
        className="form__input form__input--big"
        name="title"
        onChange={[Function]}
        type="text"
        value="fish and chips"
      />
    </label>
    <label>
      Description
      <textarea
        className="form__input form__input--big"
        name="description"
        onChange={[Function]}
        value="english street food"
      />
    </label>
    <label>
      Photo
      <input
        className="form__input"
        name="photo"
        onChange={[Function]}
        type="file"
      />
    </label>
    <div
      className="form__group"
    >
      <NewIngredient
        buttonHandler={[Function]}
        index={0}
        key="0"
        onChange={[Function]}
        value="fish"
      />
      <NewIngredient
        buttonHandler={[Function]}
        index={1}
        key="1"
        onChange={[Function]}
        value="chips"
      />
      <NewIngredient
        buttonHandler={[Function]}
        index={2}
        key="2"
        onChange={[Function]}
        value="peas"
      />
      <Button
        className="btn--primary"
        onClick={[Function]}
        text="Add New Ingredient"
        type="button"
      />
    </div>
    <div
      className="form__group"
    >
      <NewStep
        buttonHandler={[Function]}
        key="0"
        onChange={[Function]}
        value="step 1"
      />
      <NewStep
        buttonHandler={[Function]}
        key="1"
        onChange={[Function]}
        value="step 2"
      />
      <NewStep
        buttonHandler={[Function]}
        key="2"
        onChange={[Function]}
        value="step 3"
      />
      <Button
        className="btn--primary"
        onClick={[Function]}
        text="Add New Step"
        type="button"
      />
    </div>
    <button
      className="btn btn--big btn--submit"
      type="submit"
    >
      Create New Recipe
    </button>
  </form>
</div>
`;

exports[`Recipe form without data is rendering 1`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <input
      type="date"
      value="2018-09-15T00:00:00+02:00"
    />
    <label>
      Title
      <input
        className="form__input form__input--big"
        name="title"
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
    <label>
      Description
      <textarea
        className="form__input form__input--big"
        name="description"
        onChange={[Function]}
        value=""
      />
    </label>
    <label>
      Photo
      <input
        className="form__input"
        name="photo"
        onChange={[Function]}
        type="file"
      />
    </label>
    <div
      className="form__group"
    >
      <NewIngredient
        buttonHandler={[Function]}
        index={0}
        key="0"
        onChange={[Function]}
        value=""
      />
      <Button
        className="btn--primary"
        onClick={[Function]}
        text="Add New Ingredient"
        type="button"
      />
    </div>
    <div
      className="form__group"
    >
      <NewStep
        buttonHandler={[Function]}
        key="0"
        onChange={[Function]}
        value=""
      />
      <Button
        className="btn--primary"
        onClick={[Function]}
        text="Add New Step"
        type="button"
      />
    </div>
    <button
      className="btn btn--big btn--submit"
      type="submit"
    >
      Create New Recipe
    </button>
  </form>
</div>
`;

exports[`should call onSubmit prop for valid form submission 1`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <input
      type="date"
      value="2018-09-15T00:00:00+02:00"
    />
    <label>
      Title
      <input
        className="form__input form__input--big"
        name="title"
        onChange={[Function]}
        type="text"
        value="fish and chips"
      />
    </label>
    <label>
      Description
      <textarea
        className="form__input form__input--big"
        name="description"
        onChange={[Function]}
        value="english street food"
      />
    </label>
    <label>
      Photo
      <input
        className="form__input"
        name="photo"
        onChange={[Function]}
        type="file"
      />
    </label>
    <div
      className="form__group"
    >
      <NewIngredient
        buttonHandler={[Function]}
        index={0}
        key="0"
        onChange={[Function]}
        value="fish"
      />
      <NewIngredient
        buttonHandler={[Function]}
        index={1}
        key="1"
        onChange={[Function]}
        value="chips"
      />
      <NewIngredient
        buttonHandler={[Function]}
        index={2}
        key="2"
        onChange={[Function]}
        value="peas"
      />
      <Button
        className="btn--primary"
        onClick={[Function]}
        text="Add New Ingredient"
        type="button"
      />
    </div>
    <div
      className="form__group"
    >
      <NewStep
        buttonHandler={[Function]}
        key="0"
        onChange={[Function]}
        value="step 1"
      />
      <NewStep
        buttonHandler={[Function]}
        key="1"
        onChange={[Function]}
        value="step 2"
      />
      <NewStep
        buttonHandler={[Function]}
        key="2"
        onChange={[Function]}
        value="step 3"
      />
      <Button
        className="btn--primary"
        onClick={[Function]}
        text="Add New Step"
        type="button"
      />
    </div>
    <button
      className="btn btn--big btn--submit"
      type="submit"
    >
      Create New Recipe
    </button>
  </form>
</div>
`;
